### PR TITLE
Add ls cancelation support

### LINKED ISSUE #32311: Add cancelation support for LS requests
**Description:**
LSP4J provides the cancellation tokens to check whether a given request has been cancelled. Language Server can leverage this capability to avoid unnecessary compilations and computations.

**Describe your problem(s)**

**Describe your solution(s)**
Utilise the cancellation checker and pass it with the contexts and before the expensive operations, we can check whether the request has already cancelled.

### PR DESCRIPTION
**Pull request by nadeeshaan**
## Purpose
With this, we add the cancellation support for the language server operations.

Fixes #32311 

#### General
- Utilise the cancellation token provided with the LSP4J for the language server requests
- Context builders will accept the cancel checker as a parameter and will build the context.
- Each operation can access the cancel checker via the context
- Utility methods have been provided for the cancellation checks (`isCancelled`, `checkCancelled`) without accessing the cancellation token
-  **It is strongly recommended to use the utility methods in the contexts to access the semantic models, syntax tree, visible symbols. Otherwise the cancellation check has to be done explicitly when writing the feature**

#### Diagnostics
- From this point onwards, the diagnostics will not be published for each and ever text edit.
- Diagnostic publishing will be handled via a scheduler, which will be taking care of the subsequent text edits and cancel intermediate diagnostic publishing.
- Thee scheduler will publish the diagnostics after 2 sends of delay and all the diagnostic publishing requests between the particular duration will be discarded.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [x] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/util/CompletionUtil.java [hunk 1]
```diff
@@ -88,6 +89,7 @@ private CompletionUtil() {
      */
     public static List<LSCompletionItem> route(BallerinaCompletionContext ctx, Node node)
             throws LSCompletionException {
+        ctx.checkCancelled();
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/util/CompletionUtil.java:92]: Shall we add the check in the `fillTokenInfoAtCusror` method as well. As we are going to traverse the syntax tree there

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java [hunk 2]
```diff
@@ -108,12 +130,14 @@ public static SignatureContext buildSignatureContext(String uri,
                                                          WorkspaceManager workspaceManager,
                                                          SignatureHelpCapabilities capabilities,
                                                          LanguageServerContext serverContext,
-                                                         Position position) {
+                                                         Position position,
+                                                         CancelChecker cancelChecker) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:134]: Shall we update the documentation

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java [hunk 3]
```diff
@@ -129,11 +153,12 @@ public static SignatureContext buildSignatureContext(String uri,
     public static ReferencesContext buildReferencesContext(String uri,
                                                            WorkspaceManager workspaceManager,
                                                            LanguageServerContext serverContext,
-                                                           Position position) {
+                                                           Position position, CancelChecker cancelChecker) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:156]: same as above

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java [hunk 4]
```diff
@@ -149,11 +174,13 @@ public static ReferencesContext buildReferencesContext(String uri,
     public static PrepareRenameContext buildPrepareRenameContext(String uri,
                                                                  WorkspaceManager workspaceManager,
                                                                  LanguageServerContext serverContext,
-                                                                 Position position) {
+                                                                 Position position,
+                                                                 CancelChecker cancelChecker) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:178]: same as above

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java [hunk 5]
```diff
@@ -167,11 +194,13 @@ public static PrepareRenameContext buildPrepareRenameContext(String uri,
     public static RenameContext buildRenameContext(RenameParams params,
                                                    WorkspaceManager workspaceManager,
                                                    LanguageServerContext serverContext,
-                                                   LSClientCapabilities clientCapabilities) {
+                                                   LSClientCapabilities clientCapabilities,
+                                                   CancelChecker cancelChecker) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:198]: same as above

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java [hunk 6]
```diff
@@ -187,10 +216,12 @@ public static RenameContext buildRenameContext(RenameParams params,
     public static CodeActionContext buildCodeActionContext(String uri,
                                                            WorkspaceManager workspaceManager,
                                                            LanguageServerContext serverContext,
-                                                           CodeActionParams params) {
+                                                           CodeActionParams params,
+                                                           CancelChecker cancelChecker) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:220]: shall we update the documentation

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java [hunk 7]
```diff
@@ -206,11 +237,13 @@ public static CodeActionContext buildCodeActionContext(String uri,
     public static HoverContext buildHoverContext(String uri,
                                                  WorkspaceManager workspaceManager,
                                                  LanguageServerContext serverContext,
-                                                 Position position) {
+                                                 Position position,
+                                                 CancelChecker cancelChecker) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:241]: same as above

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java [hunk 8]
```diff
@@ -243,10 +276,12 @@ public static ExecuteCommandContext buildExecuteCommandContext(WorkspaceManager
      */
     public static FoldingRangeContext buildFoldingRangeContext(String uri, WorkspaceManager workspaceManager,
                                                                LanguageServerContext serverContext,
-                                                               boolean lineFoldingOnly) {
+                                                               boolean lineFoldingOnly,
+                                                               CancelChecker cancelChecker) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:280]: same as above

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java [hunk 9]
```diff
@@ -262,11 +297,13 @@ public static FoldingRangeContext buildFoldingRangeContext(String uri, Workspace
     public static BallerinaDefinitionContext buildDefinitionContext(String uri,
                                                                     WorkspaceManager workspaceManager,
                                                                     LanguageServerContext serverContext,
-                                                                    Position position) {
+                                                                    Position position,
+                                                                    CancelChecker cancelChecker) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:301]: same as above

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java [hunk 10]
```diff
@@ -280,10 +317,12 @@ public static BallerinaDefinitionContext buildDefinitionContext(String uri,
      */
     public static SemanticTokensContext buildSemanticTokensContext(String uri,
                                                                    WorkspaceManager workspaceManager,
-                                                                   LanguageServerContext serverContext) {
+                                                                   LanguageServerContext serverContext,
+                                                                   CancelChecker cancelChecker) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:321]: same as above

**Review by malinthar — CHANGES_REQUESTED** [review]
Changes look good. Have some minor change requests. I'll be trying locally to verify the expected functionality

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/util/CompletionUtil.java (hunk 1 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/util/CompletionUtil.java:92]: The `fillTokenInfoAtCusror` method internally uses the heavy operations such as retrieving the current document, which are being checked for the cancellation within the respective methods (`context.currentDocument`)

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java (hunk 10 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:335]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java (hunk 9 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:314]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java (hunk 8 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:292]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java (hunk 7 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:249]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java (hunk 6 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:227]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java (hunk 5 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:204]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java (hunk 4 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:181]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java (hunk 3 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:158]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java (hunk 2 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/ContextBuilder.java:135]: Done

**Review by IMS94 — COMMENTED** [review]
LGTM! Will test this locally and approve.

**Review by IMS94 — APPROVED** [review]
Tested locally and worked fine for me. However the 2 seconds delay + actual compile time may keep the users waiting. But that'll be fine for a start IMO.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/diagnostic/DiagnosticsHelper.java [hunk 11]
```diff
@@ -100,6 +136,47 @@ public synchronized void compileAndSendDiagnostics(ExtendedLanguageClient client
         lastDiagnosticMap.put(project.get().sourceRoot(), latestDiagnostics);
     }
 
+    /**
+     * Compiles and publishes diagnostics for a source file.
+     *
+     * @param client  Language server client
+     * @param context LS context
+     */
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/diagnostic/DiagnosticsHelper.java:144]: Missing docmentation

**Review by malinthar — APPROVED** [review]
Tested the changes locally. Works smoothly. I have just a minor comment. Please address if possible.

**Pull request merged**